### PR TITLE
Addressing PR #989 comments and resolved

### PR DESCRIPTION
**Pull request by ITZSHOAIB**
The main PR https://github.com/wevm/viem/pull/989 is closed because of the open comments that no one addressed.

- Added type to the getSocket function's argument config
- Resetting the `reconnectAttempts` after a successful connection.

---

## PR-Codex overview
This PR enhances the `getSocket` function in `rpc.ts` by adding configurable socket reconnection options and error handling.

### Detailed summary
- Added `SocketConfig` type for customizable reconnection settings
- Implemented reconnection logic with configurable options
- Added error handling for reconnection attempts
- Improved event listener setup for WebSocket events

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Review by jxom — COMMENTED** [review]
Can we please run `bun run format`? It is hard to review this diff.

**Comment by ITZSHOAIB** [comment]
> Can we please run `bun run format`? It is hard to review this diff.

Done

**Comment by 0xRapid** [comment]
Anything going on with it?

**Review by jxom — CHANGES_REQUESTED** [review]
We will need 100% test coverage for these changes.

`pnpm test:cov utils/rpc`

**Comment by nstylo** [comment]
guys, how is it with this PR? This is super important. These silent disconnects/no reconnects are impacting us in production.

**Comment by jxom** [comment]
See https://github.com/wevm/viem/pull/1894#pullrequestreview-1950039447.

**Comment by jahvi** [comment]
I've been using the fork without any issues in production, if you need it urgently I suggest you do the same until tests are covered, the change is pretty straightforward so I'm happy taking the risk.

File: src/utils/rpc.ts [hunk 1]
```diff
@@ -146,9 +146,22 @@ export type Socket = WebSocket & {
 
 const sockets = /*#__PURE__*/ new Map<string, Socket>()
 
-export async function getSocket(url_: string) {
+type SocketConfig = {
+  reconnect?: boolean
+  maxReconnectAttempts?: number
+  reconnectTimeout?: number
```
Comment (tmm) [inline src/utils/rpc.ts:152]:
```suggestion
  reconnect?: boolean | undefined
  maxReconnectAttempts?: number | undefined
  reconnectTimeout?: number | undefined
```

**Comment by jxom** [comment]
Planning on picking up this PR early next week (need to polish a few things and add tests) and merging it in.

**Comment by jxom** [comment]
Closing in favor of #2073 – your commits are preserved so you still get the contribution!

**Pull request closed without merging**